src/fs/mod.rs

```diff
@@ -118,28 +130,53 @@ impl Filesystem {
         let mut buf: Vec<u8> = vec![0; size];
         let mut total = 0;
 
-        'blocks: for block in file.blocks.iter().skip(chunk_index) {
-            let (_, mut fd) = match cache.get(block).await {
-                Ok(out) => out,
-                Err(_) => {
-                    return Ok(req.reply_error(libc::EIO)?);
-                }
+        'blocks: for block in file_metadata.blocks.iter().skip(chunk_index) {
+            // fhash works as a key inside the LRU
+            let fhash = block.hash.clone();
+            let mut lruf = self.lruf.lock().await;
+
+            // getting the file descriptor from the LRU or from the cache if not found in the LRU
+            let mut fd = if lruf.contains(&fhash) {
+                let file_descriptor = lruf.pop(&fhash).unwrap();
```

Comment:
Pop will remove the object out of the cache. Why not using Get instead? 

> NOTE: I know it's tricky because `get` will return you a reference to the file not the file itself. Hence some rework of the stored type is probably needed.

Reply:
During the time the chunk is not in the lru if there is another (async) read call at the same chunk it will have to open the file again from the cache. Right ?

Hence i still think it's better (and easier to implement) is to just keep everything in LRU (doesn't have to be last chunk) this way even if you rewind a little bit or multiple apps/users are accessing the same file it can still find some of the chunks int he lru.

Reply:
First, for your question 

> During the time the chunk is not in the lru if there is another (async) read call at the same chunk it will have to open the file again from the cache. Right ?

No, because you are using a locker inside the get function of the cache which will not permit two threads to download the same file

so that the pop function is proper than get for both sides:
- skip the reference problem 
- the block mostly will be fully downloaded so that removing it from lru will give more space to only the non-completed blocks (note that if the block is not fully downloaded, it will be added to the lru by the way)

> Hence i still think it's better (and easier to implement) is to just keep everything in LRU (doesn't ...

yes it will be easier but it will occupy the lru with a lot of completed blocks (and kick out with the same number a non-completed blocks), I think this will not be efficient

Reply:
> No, because you are using a locker inside the get function of the cache which will not permit two threads to download the same file

This is not correct. The lock is active only until you pop the file from the cache then u do `drop` of lru then another async call to read will then get a chance to run and will have access to the LRU. 
In that situation there are 2 scenarios:
- The 2nd read was trying to access a different file/chunks hence it will end up poping or opening another chunk.
- It's trying to access exactly the same chunk hence it will not find that exact chunk in LRU and then open (the same chunk) from cache 
  - this means there will be 2 open instances of the same file which might be okay, except they will race back on which file will end up in lru after it's used.

Reply:
I start to think that a GLOBAL lru cache is not the right way to go. Because the same chunk can be open by multiple read operations hence tracking the same file(plus seek position) in the same global cache is not gonna work.